### ISSUE TITLE
Use pypocketfft as FFT backend?

### ISSUE THREAD

**Issue by mreineck** [scipy/scipy#10175]
Numpy 1.17 has adopted the C version of pocketfft (https://gitlab.mpcdf.mpg.de/mtr/pocketfft) as their FFT implementation. Since then I have been working on the pocketfft code, converted it to C++ and added a Python wrapper.

The new library (https://gitlab.mpcdf.mpg.de/mtr/pypocketfft)
- supports complex and half-complex transforms, also Hartley transforms
- supports float, double and long double precision
- supports multi-dimensional FFTs
- uses CPU vector instructions for ndim>1 (i.e. it computes several 1D FFTs dimultaneously).
- uses the Bluestein algorithm for FFTs with large prime factors

Performance is significantly better than FFTPACK's, accuracy as well. For benchmarks and dicussion see https://github.com/numpy/numpy/pull/11888 and recent comments in https://github.com/numpy/numpy/issues/11885.
I have also added an interface which is very similar to `scipy.fftpack`.

Potential problems for adaptation:
- written in C++, requires a C++11 compiler and pybind11
- cosine/sine transforms are missing, and I probaly won't have the time ti implement them. They would have to be added on top of the existing machinery.

Would this be of interest for Scipy?

**Comment by ilayn** [comment]
It is certainly but on another news we are also anticipating a GSoC project for this. Don't know exactly the latest status. This effort might be combined with that.

**Comment by rgommers** [comment]
@mreineck have you followed the GSoC proposal? It just got accepted, @peterbell10 will be working on adding a backend system for FFTs to SciPy and improving the module. Peter's proposal is linked at https://mail.python.org/pipermail/scipy-dev/2019-May/023506.html, and that thread also has some discussion.

We definitely would like to get rid of FFTPACK as much as possible. I actually don't know how portable C++11 is right now - would need to look into that. If it's portable enough, we can use the new version of pypocketfft as our default implementation. Otherwise, we can use the C version that's also in numpy as our default, and enable use of the C++ version via the new backend system (which can at the same time also support pyfftw and mkl-fft).

**Comment by peterbell10** [comment]
My plan is that once backend system is in place, pocketfft would be one of the first alternate backends. From there, we could make pocketfft the default and potentially remove the fortran code altogether.

>I actually don't know how portable C++11 is right now

The [toolchain roadmap](https://scipy.github.io/devdocs/toolchain.html#c-and-c-language-standards) doesn't really say much about C++ but it sounds like some subset of C++11 is allowed. [pybind11](https://pybind11.readthedocs.io/en/stable/intro.html) only claims to support MSVC 2015 and newer though, which I suspect would be the limiting factor.

At a glance, it looks like the pocketfft code itself should work with MSVC 2013 though. We could always just bind it using cython instead of pybind11.

**Comment by ilayn** [comment]
Is there any reason why we don't directly use the C version?

**Comment by mreineck** [comment]
@rgommers I wasn't aware of the proposal, thanks for pointing it out!

One thing that could be extremely beneficial in my opinion would be some sort of unification of numpy's and scipy's FFT implementations. I'm aware that it's probably not possible to unify the interfaces (too many people are used to how things work now), but having a single implementation in the background might make maintenance a lot simpler. (This is why I provided the additional `_scipy` methods.)

**Comment by mreineck** [comment]
> Is there any reason why we don't directly use the C version?

The C version only supports double-precision 1D transforms.
Multi-dimensional transforms, different data types and vectorization were relatively easy to add using C++ templates; doing this in C would be a nightmare, unfortunately ...

**Comment by mreineck** [comment]
> At a glance, it looks like the pocketfft code itself should work with MSVC 2013 though. We could always just bind it using cython instead of pybind11.

Please let me know if you need any adjustments!

**Comment by rgommers** [comment]
> One thing that could be extremely beneficial in my opinion would be some sort of unification of numpy's and scipy's FFT implementations. I'm aware that it's probably not possible to unify the interfaces (too many people are used to how things work now), but having a single implementation in the background might make maintenance a lot simpler. (This is why I provided the additional _scipy methods.)

yes, this is definitely desirable. making the interfaces the same would be awesome, but it's indeed hard due to backwards compatibility issues. In SciPy we need to support NumPy several versions back (right now back to 1.13.3) so we can't just rely on NumPy yet. But we can take over the implementation as it is in NumPy, and then sync changes as needed.

**Comment by rgommers** [comment]
> My plan is that once backend system is in place, pocketfft would be one of the first alternate backends. From there, we could make pocketfft the default and potentially remove the fortran code altogether.

Why not make Pocketfft the default first, and then add the backend system?

> The toolchain roadmap doesn't really say much about C++ but it sounds like some subset of C++11 is allowed. pybind11 only claims to support MSVC 2015 and newer though, which I suspect would be the limiting factor.

Yes, Windows is one issue. But it may just work as you say. The other things that needs checking are more exotic platforms (ARM, AIX, etc.), and what the minimum versions of gcc/xlang/icc are that work.

@peterbell10 perhaps this is something you could look into?

**Comment by mreineck** [comment]
> The other things that needs checking are more exotic platforms (ARM, AIX, etc.)

If anyone has access to those and runs tests, it would be great if you could keep me in the loop. It should be trivial to add vectorization support there, but I couldn't do it so far because I don't know the relevant preprocessor magic on these platforms.

**Comment by WarrenWeckesser** [comment]
> making the interfaces the same would be awesome, but it's indeed hard due to backwards compatibility issues. 

Could we leave the `fftpack` module unchanged as a long-term legacy module, and create a new module, say `scipy.fft`?

The interface of `rfft` in `scipy.fftpack` is really awkward to use.  It would be nice to not have to support that in the future.

**Comment by mreineck** [comment]
> The interface of rfft in scipy.fftpack is really awkward to use.

I agree that the half-complex format is pretty strange, but it also has two massive advantages:
- since the output is real-valued and has the same dimensions as the input, in-place transforms are possible
- no obscure additional parameters are necessary for the inverse transforms to make sure you get the original array size back.

On the implementer's side, the numpy interface for real-valued transforms is a much greater pain than the scipy one ...

Of course, I prefer Hartley transforms over those two alternatives any day ;)

**Comment by WarrenWeckesser** [comment]
With a new module, the API for `rfft` could match that of numpy, and the function that returns a real array like `scipy.fftpack.rfft` could be given a new name.

**Comment by peterbell10** [comment]
>Why not make Pocketfft the default first, and then add the backend system?

Wouldn't that mean duplicating effort? First I would have to implement the current interface with pocketfft, then I would have to reimplement it against the backend interface. Depending on the design this might not be much work, but it still seems like putting the cart before the horse.

>Yes, Windows is one issue. But it may just work as you say. The other things that needs checking are more exotic platforms (ARM, AIX, etc.), and what the minimum versions of gcc/xlang/icc are that work.

I can check ARM, gcc and clang but I don't have access to AIX or icc so that would be difficult. Are minimum compiler versions documented anywhere at the moment, or at least what is used to build releases?

**Comment by rgommers** [comment]
> I can check ARM, gcc and clang but I don't have access to AIX or icc so that would be difficult.

I have access to an AIX box, I can do that. ICC we can find out fairly easily I think. I may even have a license still.

**Comment by rgommers** [comment]
> Wouldn't that mean duplicating effort? First I would have to implement the current interface with pocketfft, then I would have to reimplement it against the backend interface.

Can you clarify what you mean by "reimplement against the backend interface"? There should not be any actual implementations of anything in SciPy except for our default implementation (which is now FFTPACK, but we'd like to get rid of that because it performs badly and is hard to maintain Fortran code). So we have a new default (Pocketfft) and a backend system. The backend system will then allow users to use _external_ libraries like pyfftw or pypocketfft if those are installed.

**Comment by rgommers** [comment]
> Could we leave the fftpack module unchanged as a long-term legacy module, and create a new module, say scipy.fft?

That does sound like a good idea. It would solve the issue cleanly without any deprecations.

**Comment by larsoner** [comment]
Agreed that the `scipy.fft` namespace as the backend-switchable interface is a good way to go. We can even make `scipy.fft.*` be the same thing API-wise and underlying default implementation-wise (pocketfft) as `numpy.fft.*`. Then the difference between `numpy.fft.*` and `scipy.fft.*` is that the latter might use some other backend if you tell it to, otherwise it will do the same thing as `numpy.fft.*`.

**Comment by rgommers** [comment]
other difference perhaps that there's more transforms available, like the Hartley transform. And DCT/DST, if we're happy with the API for those we can just expose those in the new `scipy.fft`

**Comment by larsoner** [comment]
> other difference perhaps that there's more transforms available, like the Hartley transform. And DCT/DST, if we're happy with the API for those we can just expose those in the new scipy.fft

So the proposed SciPy/NumPy delination is then that `scipy.fft` is backend-able (falling back to same pocketfft interface/performance), and offers more, related transforms -- fine by me.

Regarding the half-complex `scipy.fftpack.rfft`-style interface, which I forgot to reply about -- in my experience it brings more pain and confusion than it does benefits in terms of 1) in-place operation, 2) no ambiguity in `irfft` length, and 3) no extra bytes wasted on DC imaginary component (or imaginary Nyquist part for even-length signals). So I'd rather hold off on adding that to `scipy.fft` later if people have compelling use cases for it.

Thoughts @grlee77 @endolith ?

**Comment by endolith** [comment]
Sounds like the same thing I proposed in https://github.com/scipy/scipy/issues/2487#issuecomment-255562820 :)

**Comment by peterbell10** [comment]
>Can you clarify what you mean by "reimplement against the backend interface"? There should not be any actual implementations of anything in SciPy [...]
The backend system will then allow users to use *external* libraries

In case it wasn't clear: I don't think there should be a default *and* backends; I think there should be a *default backend*. i.e. the default fft library would implement the same interface that a 3rd party backend would.

And when I say "implement", I'm strictly referring to the "glue code" that translates from the python call down to the native fft library. If the backend interface doesn't follow the exact same specification as the front end, that "glue code" might have to be rewritten. Depending on the final design this might not actually require any real effort.

>Agreed that the `scipy.fft` namespace as the backend-switchable interface is a good way to go. 

@larsoner do you mean _as well as_ `scipy.fftpack` or _instead of_? I thought that part of the idea of adding backends was that existing code using `scipy.fftpack` could benefit.

**Comment by larsoner** [comment]
I'm starting to think "instead of" (i.e., keep scipy.fftpack just for backward compatibility and focus this new effort on a new scipy.fft) might be nicer in the long term from a maintainability and usability/naming/user expectation perspective.

**Comment by rgommers** [comment]
> In case it wasn't clear: I don't think there should be a default and backends; I think there should be a default backend. i.e. the default fft library would implement the same interface that a 3rd party backend would.

Yes that makes sense. I think we're saying the same thing in different words:)

> @larsoner do you mean as well as scipy.fftpack or instead of?

I'm pretty sure "instead of". 

>  I thought that part of the idea of adding backends was that existing code using scipy.fftpack could benefit.

For most existing code it's an extra one-liner change (import `fft` instead of `fftpack`). Also, it cleans up the ugly `fftpack` name, which never made too much sense

**Comment by ilayn** [comment]
Actually `scipy.fft` would save me from a lot of troubles so +1

**Comment by endolith** [comment]
(How does it handle different backends supporting different operations, like DCT is supported by one but not another, for instance?  Will it require that every backend support every operation, even if it's a Python-level DCT↔FFT translation, or will it fallback to the pocketfft implementation in that case?)

**Comment by rgommers** [comment]
> Will it require that every backend support every operation, 

This question is also raised by the design goals in [Peter's design document](https://docs.google.com/document/d/1hGMc6ZrsmSi_LSixI5oFbo5cZkPUo92sfmnVthg1u6Y/edit#). I'll copy the two relevant ones here:

1. Any code that works with the current fftpack should work with all backends
2. Conversely, any code that doesn’t work with the current fftpack should not work with any other backends.

I'm not sure (1) is tenable, in particular because of less common functionality like DCT. What if we have one other library wanting to provide DCT, but another one doesn't? I think that should be possible. Hence I'd answer @endolith's question with "no".

I would suggest that:
- backends can provide support for any function in `scipy.fft`
- if they support it, they are responsible for providing the backend interface (the "glue") in a way that's compatible in both API and behavior. Note that that responsibility is on them though.
- using "partial implementations" of a function is probably not a good idea. conceptually a backend should say "hey I provide function X". and not "hey I don't have function X but please call X from within your Y". So regarding the *Python-level DCT↔FFT translation* question, I'd say that if a backend doesn't have DCT, then calling `dct()` should use the SciPy default for DCT and nothing else.
- it's up for discussion what should happen if a user explicitly asks for a backend and that backend doesn't provide a function that the user calls. Raise an error, or switch to the default implementation with or without a warning?

The above are for me pretty recently formed ideas based on both playing with the NumPy `__array_function__` protocol and with `uarray`. From design discussions around the former it became clear that allowing backends to provide partial implementations of the whole API is crucial to adoption. The same seems to apply here. Every backend will have `fft` and `ifft`, some may have `rfft`, and very few `dct`.

**Comment by mreineck** [comment]
> So regarding the Python-level DCT↔FFT translation question, I'd say that if a backend doesn't have DCT, then calling dct() should use the SciPy default for DCT and nothing else.

I'm not necessarily disagreeing, but you have to be aware that this can be very surprising for the user. Assume someone has a code that initially uses `rfft`; then they recognize a symmetry in their data and switch to `dct` - leading to potentially quadratic run times and worse accuracy. Exactly the opposite of what one would expect. This will be hard to explain in the documentation.

**Comment by rgommers** [comment]
@mreineck yes I agree, that's what my next point was about. A warning or exception may be clearer. I just meant, don't use half a scipy function and half a backend function from a single call to `dct`. Mixing two backends in a single function call will be the least maintainable.

**Comment by larsoner** [comment]
@rgommers's plan sounds reasonable to me, and regarding this:

> - it's up for discussion what should happen if a user explicitly asks for a backend and that backend doesn't provide a function that the user calls. Raise an error, or switch to the default implementation with or without a warning?

Somewhere we are going to set the backend using a function and/or context manager. This could be where such behavior is defined. For example if we had one of the examples from @peterbell10's doc (adapted to use the new `scipy.fft` namespace idea; whether or not it ends up this way, or using string args, does not matter too much for the sake of this discussion):
```
scipy.fft.set_backend('pyfftw')
```
We can update the signature to be by default
```
scipy.fft.set_backend('pyfftw', on_missing='fallback')
```
Where `on_missing` can be:
- `'fallback' default, meaning: silently use the built-in SciPy function, which hopefully will end up being PocketFFT anyway
- 'warn': emit a warning, then use SciPy's default
- 'raise': raise an error

This way in `'fallback'` mode you can write code that will for example try to use PyFFTW but still work if you do not have it. If on the other hand you have code that *requires* the specific backend, then you do e.g. `set_backend('cupy', on_missing='raise')` or so.

**Comment by peterbell10** [comment]
It looks like @mreineck is continuing to add features to the c++ version of pocketfft. Notably:
 * support for `long double` precision
 * multi-dimensional transforms can be multi-threading with OpenMP

The long double support doen't seem to be quite there yet. He seems to be running into issues with pybind11 (pybind/pybind11#1785) and from my testing of the c++ code, I didn't observe much precision improvement from the `long double` transforms anyway.

The multithreading on the other hand seems to be a good boost in performance.  I've tried a simple `timeit` benchmark and it gives a significant boost in performance (Although this comparison isn't very fair since pocketfft was built with native optimisations)

| Implementation | Run time |
|----------------|----------|
| pocketfft      | 21.287s  |
| pyfftw         | 48.079s  |
| fftpack        | 109.15s  |

<details>
<summary>Benchmark code</summary>

```py
import timeit

pyfftw_time = timeit.timeit(setup='''
import numpy as np
a = np.random.rand(1024*1024) + 1j * np.random.rand(1024*1024)
a = np.asarray(a, dtype=np.complex128).reshape((1024,1024))
from pyfftw.interfaces.scipy_fftpack import fftn, ifftn
''', stmt='''
args = {'threads':4, 'planner_effort':'FFTW_ESTIMATE'}
b = ifftn(fftn(a, **args), **args)
''', number=1000)
print('pyfftw: {:0.4f}s '.format(pyfftw_time))

pocketfft_time = timeit.timeit(setup='''
import numpy as np
a = np.random.rand(1024*1024) + 1j * np.random.rand(1024*1024)
a = np.asarray(a, dtype=np.complex128).reshape((1024,1024))
from pypocketfft import fftn, ifftn
''', stmt='''
b = ifftn(fftn(a, nthreads=4), nthreads=4)
''', number=1000)
print('pocketfft: {:0.4f}s '.format(pocketfft_time))

fftpack_time = timeit.timeit(setup='''
import numpy as np
a = np.random.rand(1024*1024) + 1j * np.random.rand(1024*1024)
a = np.asarray(a, dtype=np.complex128).reshape((1024,1024))
from scipy.fftpack import fftn, ifftn
''', stmt='''
b = ifftn(fftn(a))
''', number=1000)
print('FFTPACK: {:0.4f}s '.format(fftpack_time))
```

</details>

**Comment by mreineck** [comment]
> I didn't observe much precision improvement from the long double transforms anyway.

On x86_64 with the standard 80bit extended doubles, there is not much improvement to be gained... roughly a factor 1000 compared to normal doubles.
On machines with support for real 128bit floats this will be different. But I don't expect to see many people using this feature ... I added it mostly to simplify my own accuracy benchmarks.

For the moment it is a good idea to leave the scaling factor `fct` at `1.` when doing long double transforms, because, as you mention, there will be precision loss when passing the factor to C++, which negates the additional accuracy of the transforms.

> The multithreading on the other hand seems to be a good boost in performance. I've tried a simple timeit benchmark and the results look promising (Although this comparison isn't very fair since pocketfft was built with native optimisations)

I would expect that the FFTW library underneath `pyfftw` also has native optimisations on, but I might be wrong.
You will see much better `pyfftw` performance it you use its `FFTW` class directly instead of the convenience functions. In my experience, these produce considerable overhead. The naked `FFTW` calls will beat `pypocketfft` almost all of the time.

**Comment by peterbell10** [comment]
>there is not much improvement to be gained... roughly a factor 1000 compared to normal doubles.

Using your l2_error definition I was not even gaining a single order of magnitude. This was using only C++ code, avoiding the pybind issues entirely.  In fact, when I used the pybind code I couldn't get the `long double` path to execute at all.
If you're getting 3 orders of magnitude then maybe it isn't portable? Or there is something wrong with my code.

**Comment by mreineck** [comment]
Here is a quick C++ demo:

```
#include <complex>
#include <cmath>
#include <vector>
#include <iostream>
#include "pocketfft_hdronly.h"

using namespace std;
using namespace pocketfft;

template<typename T> void crand(vector<complex<T>> &v)
  {
  for (auto & i:v)
    i = complex<T>(drand48()-0.5, drand48()-0.5);
  }

template<typename T1, typename T2> long double l2err
  (const vector<T1> &v1, const vector<T2> &v2)
  {
  long double sum1=0, sum2=0;
  for (size_t i=0; i<v1.size(); ++i)
    {
    long double dr = v1[i].real()-v2[i].real(),
                di = v1[i].imag()-v2[i].imag();
    long double t1 = sqrt(dr*dr+di*di), t2 = abs(v1[i]);
    sum1 += t1*t1;
    sum2 += t2*t2;
    }
  return sqrt(sum1/sum2);
  }

template<typename T> void test(size_t len)
  {
  vector<complex<T>> data(len);
  crand(data);
  auto res=data;
  stride_t stride{sizeof(complex<T>)};
  shape_t shape{len};
  shape_t axes{0};
  c2c<T>(shape, stride, stride, axes, FORWARD,
         data.data(), res.data(), 1.);
  c2c<T>(shape, stride, stride, axes, BACKWARD,
         res.data(), res.data(), 1./len);
  cout << l2err(data,res) << endl;
  }

int main()
  {
  test<float>(1024);
  test<double>(1024);
  test<long double>(1024);
  }
```

Running this produces the following output on my computer:
```
1.53461e-07
2.82546e-16
1.45611e-19
```

**Comment by peterbell10** [comment]
Interesting, I get
```
1.53461e-07
2.82546e-16
2.74695e-16 
```

**Comment by mreineck** [comment]
Can you check `sizeof(long double)`?

**Comment by peterbell10** [comment]
```cpp
printf("%zu", sizeof(long double)); // 16
```

**Comment by mreineck** [comment]
No idea what's going on there. I would understand if you were on an older version of `pocketfft` where I had not adjusted all constants, but since you already have OpenMP support, that can't be the case.

**Comment by peterbell10** [comment]
I tested on a native linux machine and did get better precision. Apparently this is a known issue with windows subsystem for linux microsoft/WSL#830.

**Comment by grlee77** [comment]
@peterbell10
You should get a better result for pyFFTW's interfaces if you enable caching of FFTW plans. It will then also generally help to increase the planner effort to `FFTW_MEASURE`. Can you report the result you get for the following modified benchmark?

```Python
import timeit

pyfftw_time = timeit.timeit(setup='''
import numpy as np
import pyfftw
from pyfftw.interfaces.scipy_fftpack import fftn, ifftn
pyfftw.interfaces.cache.enable()
a = np.random.rand(1024*1024) + 1j * np.random.rand(1024*1024)
a = np.asarray(a, dtype=np.complex128).reshape((1024,1024))
args = {'threads':4, 'planner_effort':'FFTW_MEASURE'}
b = ifftn(fftn(a, **args), **args)  # call once to generate a plan in the cache
''', stmt='''
b = ifftn(fftn(a, **args), **args)  # call using the cached plan
''', number=1000)
print('pyfftw: {:0.4f}s '.format(pyfftw_time))
```

**Comment by grlee77** [comment]
> I would expect that the FFTW library underneath pyfftw also has native optimisations on, but I might be wrong.

If your pyFFTW is using the FFTW from conda-forge, SSE2 and AVX are enabled (AVX2 and AVX512 are not):
https://github.com/conda-forge/fftw-feedstock/blob/c30e328121c8f220a26084f5eeb546d454f27fc9/recipe/build.sh#L37

**Comment by peterbell10** [comment]
I ran both cached and uncached get 45.1541s and 49.3455s respecively. One thing I do noitice though is much higher memory consumption from pyfftw vs pypocketfft. Perhaps the GC is slowing things down.

**Comment by grlee77** [comment]
> I ran both cached and uncached with 45.1541s and 49.3455s respecively. One thing I do notice though is much higher memory consumption from pyfftw vs pypocketfft. Perhaps the GC is slowing things down.

Odd, I saw a 3x performance improvement with caching enabled (edited the previous comment with the benchmark code to list the numbers)

**Comment by peterbell10** [comment]
I'm not using conda so that might explain some of it. However, I'm a bit concerned by the memory usage: 12GB with pyfftw vs only 250MB for pocketfft.

**Comment by mreineck** [comment]
I also saw this massive memory consumption, but only with numpy 1.17 (pre-release). It went away when I switched to 1.15.4.

**Comment by mreineck** [comment]
I'm seeing the memory explosion also with numpy 1.16, but maybe it's not a question of the version number but rather an artifact of system-installed vs. pip-installed numpy package.
I'll leave that one to the experts ...

**Comment by grlee77** [comment]
> I'm not using conda so that might explain some of it. However, I'm a bit concerned by the memory usage: 12GB with pyfftw vs only 250MB for pocketfft.

I assume this high usage only with the cache enabled? I am also seeing that here (when the cache is enabled). We should open an issue at pyFFTW for that.

**Comment by grlee77** [comment]
If I run repeated, cached transforms without using `timeit`, I do not see any huge memory increase.

```Python
import numpy as np
import pyfftw
from pyfftw.interfaces.scipy_fftpack import fftn, ifftn
pyfftw.interfaces.cache.enable()
a = np.random.rand(1024*1024) + 1j * np.random.rand(1024*1024)
a = np.asarray(a, dtype=np.complex128).reshape((1024,1024))
args = {'threads':4, 'planner_effort':'FFTW_MEASURE'}
for rep in range(1000):
    b = ifftn(fftn(a, **args), **args)
```

**Comment by peterbell10** [comment]
I was on numpy 1.16.2 and downgrading to 1.15.4 also fixed the memory consumption for me. Retiming with the downgraded numpy, it goes down to 19.93s which is ~2-3x faster. 

I also find that it's timeit related somehow.

**Comment by grlee77** [comment]
Okay, it looks like `timeit` [disables garbage collection by default](https://docs.python.org/3/library/timeit.html#timeit.Timer.timeit)

If I add `gc.enable()` to the `timeit` setup string, the problem goes away for me.

So, fortunately there does not seem to be a problem in pyFFTW itself.

**Comment by mreineck** [comment]
Question for potential future development of `pypocketfft`: currently I'm making no assumptions about the strides of the passed arrays whatsoever. Would it be OK to assume that strides are always a multiple of the size of the underlying numpy data type? This will hold in all "reasonable" situations, but could be wrong if someone passes parts of a structured array, for example.

**Comment by charris** [comment]
Hmm, I think funny things can be engineered, although IIRC, we routinely divide strides by itemsize to get sizes. @seberg Thoughts?

**Comment by seberg** [comment]
Not sure that dividing strides by itemsize is much of a thing, dividing shape by itemsize for contiguous data is a thing though.
Yes, it will hold on most reasonable data, but not on all supported data. Its similar to assuming the data is aligned maybe. Likely you can get away with it by simply falling back to a slow "copy" path for the hopefully rare case where it cannot be done.

**Comment by mreineck** [comment]
> Its similar to assuming the data is aligned maybe.

That's an assumption I make implicitly. If array entries are not sufficiently aligned (I'm not talking about sufficient alignment for vectors here, just about the natural alignment required for the data type) the module will crash.
Unfortunately requiring that strides are divisible by itemsize is a much stronger requirement. `complex<double>` only requires 8-byte alignment normally, although its size is 16. For `long double`, discrepancies are probably even larger.

> Likely you can get away with it by simply falling back to a slow "copy" path for the hopefully rare case where it cannot be done.

It's not about avoiding copying (I'm doing this already where possible). It's just about my multi-D array class on the C++ side, which does all offset calculations in bytes and casts to the correct data type every time I ask for an element. This is pretty awkward, and I don't know whether it inhibits some optimizations.

Anyway, if we cannot exclude weird strides, I'll just leave things as they are. Thanks for your feedback!

**Comment by aarchiba** [comment]
> That's an assumption I make implicitly. If array entries are not sufficiently aligned (I'm not talking about sufficient alignment for vectors here, just about the natural alignment required for the data type) the module will crash.
Just to check: what alignment do you assume for 80-bit floats? Numpy offers both 12-byte and 16-byte alignment. 

One can, of course, create arrays whose data items are not aligned; this is rare enough that a copy on input is fine, but a segfault is not.

**Comment by seberg** [comment]
@mreineck but you are then talking about purely numeric arrays, which would be a copy of an old array if necessary? In that case I think you are covered by contiguity (even if it may be along a different axis), so I *think* you are fine, although I am not quite sure I follow.

**Comment by peterbell10** [comment]
Is it worth handling views into structured array that lack natural alignment:
```python
>>> np.zeros(10, dtype='u1,f8')['f1'].strides
(9,)
```
It sounds like pypocketfft would crash on that currently.

**Comment by mreineck** [comment]
> Just to check: what alignment do you assume for 80-bit floats? Numpy offers both 12-byte and 16-byte alignment.

I'm assuming that the data is aligned in such a way that the CPU can read an object of that type from the address, that's all.

> One can, of course, create arrays whose data items are not aligned; this is rare enough that a copy on input is fine, but a segfault is not.

If I'm on a machine that requires `double`s to be 8-byte aligned and I'm passed a `double *` pointing to an adress that is not divisible by 8, than a segfault is perfectly reasonable. Such a pointer invokes undefined behaviour. Does `numpy` really do such things?

**Comment by mreineck** [comment]
```
>>> np.zeros(10, dtype='u1,f8')['f1'].strides
(9,)
```
> It sounds like pypocketfft would crash on that currently.

It seems that this kind of structure works on `x86` (at a performance penalty). It wouldn't work on ARM CPUs, as far as I understand. If `numpy` produces such stredes there, that's a bug.

**Comment by peterbell10** [comment]
> this kind of structure [...] wouldn't work on ARM CPUs

It's not that it doesn't work on arm, it's just that you have to `memcpy` into aligned storage instead of just casting to double.

https://gcc.godbolt.org/z/TNnNWJ

**Comment by mreineck** [comment]
> It's not that it doesn't work on arm, it's just that you have to memcpy into aligned storage instead of just casting to double.

Given the above example with 9-byte strides, I would have to call an 8-byte `memcpy` for every single entry in the array. That's not really nice performance-wise.
The problem is that I do not know a priori on which hardware this is necessary, so the only safe solution would be to this copy everytime I detect odd alignment and/or strides...

Could you please point me to where this is done for `fftpack`, for example?

**Comment by seberg** [comment]
@mreineck numpy arrays have some aligned flags (and a bit more annoying in depth handling) to detect that this may be possible. And yes, we do fall back to slow paths in that case, which for example copy each element first (not typically with memcpy, but I think that is mostly historic). Numpy does support for example Sparc, and such alignment issues cause bus errors for them, so it is a real issue (if users create funny enough arrays). But, I do feel that you should be able to just check if the numpy array is "behaved" (although that includes writeable), and if not create a copy.

**Comment by aarchiba** [comment]
> > Just to check: what alignment do you assume for 80-bit floats? Numpy offers both 12-byte and 16-byte alignment.

> I'm assuming that the data is aligned in such a way that the CPU can read an object of that type from the address, that's all.

It sounds like on x86 this isn't a problem (except for speed), and by default numpy chooses to pack 10-byte floats into 16-byte spaces if on a 64-bit architecture, but users can specify float96 to get 12-byte floats on any architecture, and they might in an attempt to save space.

Numpy is sometimes used on memory-mapped files, so occasionally an array arises in some layout optimized for storage on disk rather than efficient CPU operations. It is important that numpy be able to read and write such arrays in place, but it is understood that operations on such arrays might be slower and/or require a copy compared to more conventional layouts. In fact many numpy operations (for example reshape) share memory only on a best-effort basis, copying when necessary, in particular when a library needs a particular kind of layout. So it seems to me perfectly reasonable to specify your layout requirements up front and call something like np.ascontiguousarray if they are not met. (In fact an FFT might be faster on the copy than the original even for some normally strided arrays depending on cache layout and operation order.) For something as basic as an FFT, it would be nice to avoid this copy for most use cases, but people with funny alignments are expecting slowness and copies. Segfaults are not okay.

**Comment by peterbell10** [comment]
>The problem is that I do not know a priori on which hardware this is necessary, so the only safe solution would be to this copy everytime I detect odd alignment and/or strides...

If you look at the [godbolt link](https://gcc.godbolt.org/z/TNnNWJ), gcc on intel turns the memcpy into a `movsd` even on `-O0`.

**Comment by mreineck** [comment]
That's great ... but on ARM this won't happen, so if I don't want to penalize users on that platform, I can't avoid having two code paths.

I'd like to address this outside of `pocketfft` (i.e. somewhere in `pypocketfft.cc` instead). Unfortunately this means (potentially) copying the entire multi-D arrays (`out` may also need to be copied at exit), which is a big memory overhead.

For the current `fftpack` implementation, is the aligning done somewhere in the automatically generated Fortran wrapper? I tried with the 9-byte-strided test array, and it was not copied before calling the wrapper, so it must be in there somewhere...

**Comment by mreineck** [comment]
OK, here is the dilemma I'm facing:

Assume someone passes an array to `pypocketfft` which doesn't have the `aligned` flag set. If I make an aligned copy the array, I'm wasting space on platforms where this is not necessary (and could lead to an out-of-memory condition); if I don't copy, the code will crash on some CPUs. So I would have to make a distinction based on host CPU type, which is an information I do not have.

I do not want to change `pocketfft_hdronly.h` itself, since by definition of its interface it cannot be passed misaligned pointers (I'm sure that a `double *` from which it is impossible to read a `double` is ill-formed C++). So something must be done in `pypocketfft.cc`, and I have no idea what :-/

BTW, `pyfftw` refuses to create plans for unaligned and strangely strided input/output arrays. That would also be an acceptable solution for me, but if I understood correctly this is not how `scipy` does things.

**Comment by peterbell10** [comment]
What if you just used `ndarr` in the interface to `pocketfft_hdronly.h`? The non-contiguous case in `general_c` already does a copy so changing that to `memcpy` shouldn't be any more work.

**Comment by mreineck** [comment]
As you showed, using `memcpy` for small fixed sizes works fine on some platforms like x86, because the compiler knows it can simply replace the `memcpy` with an unaligned load/store. But your demo code on godbolt also showed that on ARM, `memcpy` is called unconditionally, which must be a big performance hit. This could only be avoided by another code path, which feels like overkill for a scenario that should never occur in a performant numerical library. 
To be honest, even the existing `char *` hackery in `ndarr` feels scary to me now, and I'd prefer to go back to strides measured in sizes of the underlying data types instead of bytes.
If we fix this, please let's fix it in the wrapper ... perhaps even outside of `pypocketfft.cc`. I'm pretty sure that most other FFT backends won't be able to deal with weirdly aligned/strided arrays either, so the alignment code will certainly be re-used.

**Comment by peterbell10** [comment]
I think that's fine too. I'd guess this is a very rare edge case so I wouldn't be too woried about just copying the whole array. Especially since the resulting transform could usually be done in place so it wouldn't necessarily have any memory overhead.

**Comment by mreineck** [comment]
> Especially since the resulting transform could usually be done in place so it wouldn't necessarily have any memory overhead.

Hah, very good point! Now I don't feel so bad any more :)

**Comment by aarchiba** [comment]
> Assume someone passes an array to pypocketfft which doesn't have the aligned flag set. If I make an aligned copy the array, I'm wasting space on platforms where this is not necessary (and could lead to an out-of-memory condition); if I don't copy, the code will crash on some CPUs. So I would have to make a distinction based on host CPU type, which is an information I do not have.

Because this is a rare case, the main concern is avoiding crashes or "I can't do that" exceptions; extra slowness or memory use are a tolerable cost. (People whose arrays are pushing the limits of virtual memory can be required to be aware of alignment issues.) So when given a non-aligned array, why not make the copy unless you know it's going to be okay? That is, the code path contains exceptions to "make a copy if non-aligned" for Intel and other architectures explicitly known to be okay? 

I agree this should live in the wrapper. I guess the question is, can pypocketfft live with byte-based strides? If so, then the wrapper can skip the copy when it knows that's safe. (And when a copy is made, the code can be permitted to use in-place FFTs, if those are faster.)

**Comment by mreineck** [comment]
> So when given a non-aligned array, why not make the copy unless you know it's going to be okay? That is, the code path contains exceptions to "make a copy if non-aligned" for Intel and other architectures explicitly known to be okay?

Just to be sure, where do you suggest these tests should happen - in Python or already on the C++ side? Personally, I'd prefer Python, but I don't know how well the computing environment can be queried from there.

> I guess the question is, can pypocketfft live with byte-based strides?

Yes, I can keep this functionality in. "Natural" strides look a bit nicer, but it's not a dramatic difference.

> (And when a copy is made, the code can be permitted to use in-place FFTs, if those are faster.)

They'll just be more economic in terms of memory, but probably not noticeably faster.

**Comment by aarchiba** [comment]
> > So when given a non-aligned array, why not make the copy unless you know it's going to be okay? That is, the code path contains exceptions to "make a copy if non-aligned" for Intel and other architectures explicitly known to be okay?

> Just to be sure, where do you suggest these tests should happen - in Python or already on the C++ side? Personally, I'd prefer Python, but I don't know how well the computing environment can be queried from there.

I'd say put the tests wherever convenient. Since there are already python-level functions to produce array copies when necessary, python might be easier. I don't know how easy these architecture checks are from python either, but if absolutely necessary you could just add a "required_alignment" function to the C++ side that the python can call to check. Ideally numpy would have a function that took an array and copied it if necessary to produce an array where element access from C++ didn't cause a segfault, but numpy.require expects you to know whether you need aligned elements.
